sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/DeltaWriter.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.write;
+
+import java.io.IOException;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * A data writer returned by {@link DeltaWriterFactory#createWriter(int, long)} and is
+ * responsible for writing a delta of rows.
+ *
+ * @since 3.4.0
+ */
+@Experimental
+public interface DeltaWriter<T> extends DataWriter<T> {
```

Comment:
One more comment: We use the type parameter T because it can be in either row or columnar format. However, I think this new delta writer can only work with rows. We should probably do `DeltaWriter extends DataWriter<InternalRow>`

Reply:
Let me take a closer look on Monday.

Reply:
@cloud-fan, could you elaborate a bit on why you think we can only work with rows here?

Reply:
because we have `void delete(T metadata, T id) throws IOException;`. Are we going to perform a delete with a batch of metadata and id rows?

Reply:
Well, I can see us passing batches of deletes and metadata at some point in the future. We can assume values at the same index will belong to the same row.